dataset/src/main/scala/frameless/macros/ColumnMacros.scala

```diff
@@ -0,0 +1,377 @@
+        } else addError(argsE)(tree, "Only constructor can be used here")
+      }
+      // Constructing a tuple or parameterized case class
+      case Apply(ta @ TypeApply(sel @ Select(qualifier, TermName("apply")), typArgs), AllExprs(argsE)) => Some {
+        if(isConstructor(ta, tree.tpe)) {
+          argsE.right.map {
+            args =>
+              val params = sel.symbol.asMethod.paramLists.head
+              val names = sel.symbol.asMethod.paramLists.head.zip(args).map {
+                case (param, paramExpr) => param.name.toString -> paramExpr
+              }.toMap
+              Construct(tree.tpe, names)
+          }
+        } else addError(argsE)(tree, "Only constructor can be used here")
+      }
+
+      // A binary operator - the operator must exist on org.apache.spark.sql.Column
+      case Apply(sel @ Select(This(lhsE), op: TermName), List(This(rhsE))) => Some {
+        subBinaryOperators.get(sel.symbol.asMethod).map {
+          sub => requireBoth(lhsE, rhsE)(
+            (lhs, rhs) => failFrom(lhs, rhs),
+            (lhs, rhs) => Right(TrustMeBro(tree.tpe, sub(Expr.refold(lhs), Expr.refold(rhs))))
+          )
+        }.getOrElse {
+          if (isColumnOp(op, 1)) {
+            requireBoth(lhsE, rhsE)(
+              (lhs, rhs) => failFrom(lhs, rhs),
+              (lhs, rhs) => Right(BinaryOperator(tree.tpe, lhs, op, rhs))
+            )
+          } else addError(lhsE, rhsE)(tree, s"${op.decodedName} is not a valid column operator")
+        }
+      }
+
+      // A function application - what to do with this? How can we check if it's an OK function?
+      // Check org.apache.spark.sql.functions?
+      // I think we have to port all spark functions to typed versions so we can typecheck here
+      case Apply(Select(src, fn: TermName), AllExprs(argsE)) =>
+        Some(Left(List((tree, "Function application not currently supported"))))
+        //Some(argsE.right.map(args => FunctionApplication(tree.tpe, src, fn, args)))
+
+      case Apply(_, _) => Some(Left(List((tree, "Function application not currently supported"))))
+
+      case _ => None
+    }
+
+    private def isColumnOp(name: TermName, numArgs: Int): Boolean = {
+      val sym = weakTypeOf[org.apache.spark.sql.Column].member(name)
+      sym.isMethod && {
+        val bothEmptyArgs = numArgs == 0 && sym.asMethod.paramLists.isEmpty
+        val matchingArgCounts = sym.asMethod.paramLists.map(_.length) == List(numArgs)
+        bothEmptyArgs || matchingArgCounts
+      }
+    }
+
+    def isConstructor(tree: Tree, result: Type, appliedTypes: Option[List[Type]] = None): Boolean = {
+      val (qualifier, typeArgs) = tree match {
+        case TypeApply(Select(q, _), args) => (q, args)
+        case Select(q, _) => (q, Nil)
+      }
+
+      val MethodType(params, _) = tree.tpe
+
+      val companion = Option(result.companion).filterNot(_ == NoType).getOrElse {
+        try {
+          c.typecheck(Ident(result.typeSymbol.name.toTermName)).tpe
```

Comment:
This is for the case when the case class is nested inside a class or function - you can't directly get its companion. But I think that cases like that won't work anyway (breaks Spark's codegen), so maybe this won't be needed.